src/utils/routerTradeAdapter.ts

```diff
@@ -0,0 +1,232 @@
+import { MixedRouteSDK, Trade as RouterTrade } from '@uniswap/router-sdk'
+import {
+  ChainId,
+  Currency,
+  CurrencyAmount,
+  Ether,
+  NativeCurrency as SdkNativeCurrency,
+  Token,
+  TradeType,
+} from '@uniswap/sdk-core'
+import { Pair, Route as V2Route } from '@uniswap/v2-sdk'
+import { Pool, Route as V3Route, FeeAmount } from '@uniswap/v3-sdk'
+import { BigNumber } from 'ethers'
+import { ETH_ADDRESS, WRAPPED_NATIVE_CURRENCY } from './constants'
+
+export type TokenInRoute = {
+  address: string
+  chainId: number
+  symbol: string
+  decimals: string
+  name?: string
+  buyFeeBps?: string
+  sellFeeBps?: string
+}
+
+export enum PoolType {
+  V2Pool = 'v2-pool',
+  V3Pool = 'v3-pool',
+}
+
+export type V2Reserve = {
+  token: TokenInRoute
+  quotient: string
+}
+
+export type V2PoolInRoute = {
+  type: PoolType.V2Pool
+  address?: string
+  tokenIn: TokenInRoute
+  tokenOut: TokenInRoute
+  reserve0: V2Reserve
+  reserve1: V2Reserve
+  amountIn?: string
+  amountOut?: string
+}
+
+export type V3PoolInRoute = {
+  type: PoolType.V3Pool
+  address?: string
+  tokenIn: TokenInRoute
+  tokenOut: TokenInRoute
+  sqrtRatioX96: string
+  liquidity: string
+  tickCurrent: string
+  fee: string
+  amountIn?: string
+  amountOut?: string
+}
+
+export type PartialClassicQuote = {
+  // We need tokenIn/Out to support native currency
+  tokenIn: string
+  tokenOut: string
+  tradeType: TradeType
+  route: Array<(V3PoolInRoute | V2PoolInRoute)[]>
+}
+
+interface RouteResult {
+  routev3: V3Route<Currency, Currency> | null
+  routev2: V2Route<Currency, Currency> | null
+  mixedRoute: MixedRouteSDK<Currency, Currency> | null
+  inputAmount: CurrencyAmount<Currency>
+  outputAmount: CurrencyAmount<Currency>
+}
+
+export const isNativeCurrency = (address: string) => address.toLowerCase() === ETH_ADDRESS.toLowerCase()
+
+// Helper class to convert routing-specific quote entities to RouterTrade entities
+// the returned RouterTrade can then be used to build the UniswapTrade entity in this package
+export class RouterTradeAdapter {
+  // Generate a RouterTrade using fields from a classic quote response
+  static fromClassicQuote(quote: PartialClassicQuote) {
+    const { route, tokenIn, tokenOut } = quote
+
+    const tokenInData = route[0]?.[0]?.tokenIn
+    const tokenOutData = route[0]?.[route[0]?.length - 1]?.tokenOut
```

Comment:
think it would be cleaner to do more of this validation ahead of time instead of with weak nulls, like

```
if (route.length === 0) throw
if (route[0].length === 0) throw
const first Path = route[0][0];
```

or something -- just best to add more clarity especially since now this is depended on by many more services

Reply:
Yes good call out - this code was mostly copied over from other impl to ensure correctness but I agree